snuba/subscriptions/scheduler_processing_strategy.py

```diff
@@ -0,0 +1,150 @@
+    tick.
+
+    `max_ticks_buffered_per_partition` applies if the scheduler mode is
+    WAIT_FOR_SLOWEST_PARTITION. Once the maximum ticks is received for that
+    partition, we start to submit ticks for processing even if that timestamp
+    is not received for all partitions yet.
+
+    This prevents the buffered tick list growing infinitely and scheduling
+    to grind to a halt if one partition starts falling far behind for some reason.
+    """
+
+    def __init__(
+        self,
+        mode: SchedulerMode,
+        partitions: int,
+        max_ticks_buffered_per_partition: Optional[int],
+        next_step: ProcessingStrategy[Tick],
+        metrics: MetricsBackend,
+    ) -> None:
+        if mode == SchedulerMode.WAIT_FOR_SLOWEST_PARTITION:
+            assert max_ticks_buffered_per_partition is not None
+
+        self.__mode = mode
+        self.__partitions = partitions
+        self.__max_ticks_buffered_per_partition = max_ticks_buffered_per_partition
+        self.__next_step = next_step
+        self.__metrics = metrics
+
+        self.__buffers: Mapping[int, Deque[Message[Tick]]] = {
+            index: deque() for index in range(self.__partitions)
+        }
+
+        # Stores the latest timestamp we received for any partition. This is
+        # just for recording the partition lag.
+        self.__latest_ts: Optional[datetime] = None
+
+    def poll(self) -> None:
+        pass
+
+    def submit(self, message: Message[Tick]) -> None:
+        # If the scheduler mode is immediate or there is only one partition
+        # immediately submit message to the next step.
+        # We don't keep any latest_ts values as it is not relevant.
+        if self.__mode == SchedulerMode.IMMEDIATE or self.__partitions == 1:
+            self.__next_step.submit(message)
+            return
+
+        # Update the latest_ts for metrics
+        if (
+            self.__latest_ts is None
+            or message.payload.timestamps.upper > self.__latest_ts
+        ):
+            self.__latest_ts = message.payload.timestamps.upper
+
+        tick_partition = message.payload.partition
+        self.__buffers[tick_partition].append(message)
+
+        # If the buffer length exceeds `max_ticks_buffered_per_partition`
+        # immediately submit the earliest message in that buffer to the next step.
+        if len(self.__buffers[tick_partition]) > cast(
+            int, self.__max_ticks_buffered_per_partition
+        ):
+            self.__next_step.submit(self.__buffers[tick_partition].popleft())
+            return
+
+        # If there are any empty buffers, we can't submit anything yet.
+        # Otherwise if all the buffers have ticks then we look for the partition/s
+        # with the earliest tick (i.e. the tick with the earliest upper timestamp
+        # interval value) and submit it to the next step.
+        if len(self.__buffers[tick_partition]) > 1:
+            return
+
+        while True:
```

Comment:
Can't you replace this `while True:` (which always makes me nervous on a loop that has nested loops and that is fairly long) with a loop like this:
```
# Keep doing it as long as there is a tick per partition in the buffer.
while all(len(buffer) > 0 for buffer in self.__buffers.values):
     # populate earliest_ts_partitions
     # send the ticks in earliest_ts_paritions
     # record metrics
```

It will be shorter and won;t need break nor continue

Reply:
Perhaps this is overthinking things but I deliberately did `while True` in order to avoid the unnecessary check of the 64 (well... 64 in our specific case right now) buffers on each iteration of this loop. I made a change to get rid of some of the earlier break/continue stuff that was a bit confusing. Not sure the current implementation makes you any less nervous though.

Reply:
I would not freak out about 64 (or even 128 or 256) iterations over the buffer. That is not going to be your bottleneck in this consumer by a long shot as you will have Kafka consumption, production, message serialization and generation of a message per subscription (which are 2 orders of magnitude more).